### ISSUE TITLE
files.exclude isn't applied to fileCreation/Deletion/Changed

### ISSUE THREAD

**Issue by sean-mcmanus** [microsoft/vscode-cpptools#5141]
@sean-mcmanus Thanks, it has installed the new version. Whilst building I can see:
cpptools/fileCreated: file:///<folder root>/.git/lfs/tmp/065377983
I thought everything in the .git folder would be excluded.
I can also see some files specific to our setup which I should be able to exclude. I've got a directory "a/b" in the "Files: Exclude", but cpptools is stilll seeing files in it. This directory is where our build system stores a lot of temporary files.

_Originally posted by @BillDenton in https://github.com/microsoft/vscode-cpptools/issues/4741#issuecomment-600254700_

**Comment by sean-mcmanus** [comment]
We check files.exclude in regards to what files get added to our database, but there is other unnecessary processing that is still occurring for files matching files.excludes.

We may also want to handle files.watcherExclude...there might be another issue tracking that already.

**Comment by sean-mcmanus** [comment]
@BillDenton Can you let us know how severe the CPU impact is in your scenario? From the code, it looks like some cases could be problematic.

**Comment by BillDenton** [comment]
@sean-mcmanus On both my Windows 10 PCs on all folders (using storing same large code base on different git branches) the C/C++ extension always (as far as I know) uses a whole CPU core all the time. It doesn't matter if I'm building or not or editing a file. We have discussed this another issue. If I remember correctly it was connected the extension handling large codebases. Other people working on the same codebase have the same issue - possibly not everyone who uses the C/C++ extension.

**Comment by sean-mcmanus** [comment]
I don't think we have tracked down a cause/repro for the CPU usage "all the time" issue. However, we did find/fix a major problem that could occur after switching branches in which our parse queues could get backlogged with huge amounts of (potentially duplicate) work: https://github.com/microsoft/vscode-cpptools/issues/4894 .

**Comment by BillDenton** [comment]
I've just checked the logs again still seeing stuff like:
cpptools/fileCreated: file:///<redacted>/.git/lfs/tmp/798668719
that was during git pull
Also lots of other temp excluded files in my repo.
Any update?

**Comment by sean-mcmanus** [comment]
This hasn't been fixed yet.

**Comment by BillDenton** [comment]
@sean-mcmanus Any update? I'm on 0.29.0 and each VS Code window usually takes out a physical core on my 6 core Windows laptop. I notice that there are quite a few other performance issues, for example, #5678, #5574, #4418,

**Comment by sean-mcmanus** [comment]
Unfortunately, we haven't had a chance to investigate this further yet.

**Comment by BillDenton** [comment]
@sean-mcmanus @bobbrow Any update? Still using a core per VS Code instance. Still seeing "cpptools/fileDeleted" or "cpptools/fileCreated" for files that are in directories that excluded by files.exclude. All our intermediate ".obj" files are stored in "build" directories in many places across our source tree. I tried all of the following to exclude them:
   "files.exclude": {
<other excludes>
      "**/build": true,
      "*app*/**/build": true
      "build/": true,
}
I'm not sure how to exclude all directories called "build" in all parts of the directory tree.

**Comment by sean-mcmanus** [comment]
The fix should be available in our next release (potentially in a week or two, either as 1.2.0-preview2 or 1.2.0-insiders).

The `"**/build": true" should exclude all the build folders. Our code wasn't checking files.excludes for the file watcher messages.

**Comment by sean-mcmanus** [comment]
Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/1.2.0-insiders2

**Comment by BillDenton** [comment]
@sean-mcmanus Updated to the insiders build and I can't see it looking at excluded files (in directories named "build"). So looks better, but it still uses a whole core on my laptop all the time. Other people are also seeing the same problem. Almost as efficient as mining bitcoin :-)

**Comment by sean-mcmanus** [comment]
@BillDenton Was there an issue tracking that remaining issue? Did we ever figure out what processing is occurring, i.e. a call stack?

**Comment by BillDenton** [comment]
@sean-mcmanus I think we last discussed it in https://github.com/microsoft/vscode-cpptools/issues/4418. This problem has been seen for well over a year (I can't remember when it started). Other people also see it working on the same code base (I don't know if it is everyone as a lot of people use VS Code).

**Comment by BillDenton** [comment]
@sean-mcmanus Here are some example stack traces from SysInternals Process Explorer provided by @StephenHillAtViaivi.
ntoskrnl.exe!KeSynchronizeExecution+0x5b46
ntoskrnl.exe!KeWaitForSingleObject+0x1c2d
ntoskrnl.exe!KeWaitForSingleObject+0xab4
ntoskrnl.exe!KeWaitForSingleObject+0x255
ntoskrnl.exe!IoApplyPriorityInfoThread+0x12d7
ntoskrnl.exe!KeWaitForSingleObject+0x3828
ntoskrnl.exe!KiCheckForKernelApcDelivery+0x2b
ntoskrnl.exe!KeLeaveCriticalRegion+0x37
ntoskrnl.exe!ExEnterCriticalRegionAndAcquireResourceShared+0x125
ntoskrnl.exe!NtCreateEvent+0x8d5
ntoskrnl.exe!ObOpenObjectByNameEx+0xfe8
ntoskrnl.exe!NtClose+0xde
ntoskrnl.exe!_setjmpex+0x7bb5
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x544
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x503
wow64cpu.dll!BTCpuSimulate+0x9
wow64.dll!Wow64LdrpInitialize+0x26a
wow64.dll!Wow64LdrpInitialize+0x127
ntdll.dll!LdrInitializeThunk+0x17b
ntdll.dll!LdrInitializeThunk+0x63
ntdll.dll!LdrInitializeThunk+0xe
0x0000000000000000


ntoskrnl.exe!KeSynchronizeExecution+0x5b46
ntoskrnl.exe!KeWaitForSingleObject+0x1c2d
ntoskrnl.exe!KeWaitForSingleObject+0xab4
ntoskrnl.exe!KeWaitForSingleObject+0x255
ntoskrnl.exe!IoApplyPriorityInfoThread+0x12d7
ntoskrnl.exe!KeWaitForSingleObject+0x3828
ntoskrnl.exe!KiCheckForKernelApcDelivery+0x2b
ntoskrnl.exe!KeLeaveGuardedRegion+0x2c
FLTMGR.SYS!FltRetrieveFileInfoOnCreateCompletion+0x5db
FLTMGR.SYS!FltRetrieveFileInfoOnCreateCompletion+0x3b7
FLTMGR.SYS!FltDecodeParameters+0x225a
FLTMGR.SYS!FltDecodeParameters+0x1840
FLTMGR.SYS!FltDecodeParameters+0x13b2
FLTMGR.SYS!FltDecodeParameters+0x119e
ntoskrnl.exe!IofCallDriver+0x59
ntoskrnl.exe!ProbeForWrite+0x19d
ntoskrnl.exe!NtFreeVirtualMemory+0x8f0
ntoskrnl.exe!ObfDereferenceObjectWithTag+0xc9
ntoskrnl.exe!ObOpenObjectByNameEx+0x1037
ntoskrnl.exe!NtClose+0xde
ntoskrnl.exe!_setjmpex+0x7bb5
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x544
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x503
wow64cpu.dll!BTCpuSimulate+0x9
wow64.dll!Wow64LdrpInitialize+0x26a
wow64.dll!Wow64LdrpInitialize+0x127
ntdll.dll!LdrInitializeThunk+0x17b
ntdll.dll!LdrInitializeThunk+0x63
ntdll.dll!LdrInitializeThunk+0xe


[11:17] Stephen Hill
    ntoskrnl.exe!KeSynchronizeExecution+0x5b46
ntoskrnl.exe!KeWaitForSingleObject+0x1c2d
ntoskrnl.exe!KeWaitForSingleObject+0xab4
ntoskrnl.exe!KeWaitForSingleObject+0x255
ntoskrnl.exe!IoApplyPriorityInfoThread+0x12d7
ntoskrnl.exe!KeWaitForSingleObject+0x3828
ntoskrnl.exe!KiCheckForKernelApcDelivery+0x2b
ntoskrnl.exe!KeLeaveCriticalRegion+0x37
Ntfs.sys+0x11868b
ntoskrnl.exe!IofCallDriver+0x59
FLTMGR.SYS!FltIsCallbackDataDirty+0x2ee
FLTMGR.SYS!FltQueryInformationFile+0x977
ntoskrnl.exe!IofCallDriver+0x59
ntoskrnl.exe!PsGetProcessWow64Process+0x104
ntoskrnl.exe!SeQueryInformationToken+0x22be
ntoskrnl.exe!SeUnlockSubjectContext+0x85f
ntoskrnl.exe!ObOpenObjectByNameEx+0x201
ntoskrnl.exe!NtCreateFile+0x696
ntoskrnl.exe!NtOpenFile+0x58
ntoskrnl.exe!_setjmpex+0x7bb5
ntdll.dll!ZwOpenFile+0x14
wow64.dll!Wow64ShallowThunkSIZE_T64TO32+0x204a
wow64.dll!Wow64SystemServiceEx+0x153
wow64cpu.dll!TurboDispatchJumpAddressEnd+0xb
wow64cpu.dll!BTCpuSimulate+0x9
wow64.dll!Wow64LdrpInitialize+0x26a
wow64.dll!Wow64LdrpInitialize+0x127
ntdll.dll!LdrInitializeThunk+0x17b
ntdll.dll!LdrInitializeThunk+0x63
ntdll.dll!LdrInitializeThunk+0xe

**Comment by sean-mcmanus** [comment]
Those stack traces don't make any sense to me and don't show any of our code's stack frames. Are you able to attach a debugger and just pause it (see https://github.com/microsoft/vscode-cpptools/wiki/Attaching-debugger-to-cpptools-or-cpptools%E2%80%90srv ). You should see lots of threads with small call stack frames that a not doing work that you can ignore and then 1 or a few threads that have bigger call stacks that appear to be doing real work.

**Comment by BillDenton** [comment]
@sean-mcmanus I'm starting to capture the stack. I'll update this comment with the details.
I've got one VS Code window open and can see in Task Manager:
![image](https://user-images.githubusercontent.com/17859340/105983185-3d9a1f80-6090-11eb-8e07-d01956b3e4c5.png)

**Comment by sean-mcmanus** [comment]
The CPU usage is due to one thread (the others have no work):
```
KernelBase.dll!FindFirstFileExW() (Unknown Source:0)
cpptools.exe!msvc::enumerate_files_win32(char const *,class std::deque<class msvc::directory_utf8_entry,class std::allocator > &,int) (Unknown Source:0)
```

That code is just enumerating files in browse.paths (see https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirstfileexw and https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findnextfilew). It's normally supposed to be fast (no parsing is done) unless it includes some path that has millions of files or a symlink to such a place. We have an issue tracking enabling a files.exclude workaround at https://github.com/microsoft/vscode-cpptools/issues/3123 . The current workaround would be to somehow remove the paths from your browse.path setting, possibly replacing them with paths with less child paths. Are you able to determine why this would be slow in your scenario?

**Comment by BillDenton** [comment]
@sean-mcmanus By browse.path do you mean in .vscode\c_cpp_properties.json under "name": "Win32"? At the moment there is nothing in that section.

**Comment by sean-mcmanus** [comment]
We do stuff under the hood to populate it. The includePath gets copied to there and also the system defines from the compilerPath. You can explicitly set it to avoid having the includePath copied, but you'd need to set the compilerPath to "" to avoid the system includes getting copied (we don't have a way to bypass that), but my guess is that you're not hitting a problem with system includes (which tends to be more of a potential issue on Linux/Mac). Also, using a path in browse.path that ends with "/*" makes it non-recursive.

**Comment by sean-mcmanus** [comment]
We could change it to do a breadth first traversal instead of depth first and then add log messages when it goes a new depth level and then abort after some time like 4? minutes to unblock users who are hitting this unusual case. For most users it takes under a minute.

**Comment by BillDenton** [comment]
@sean-mcmanus It does seem like an abort message with a timeout (from a setting) would be a good idea. I tried:
"compilerPath": "",
            "browse": {
                "path": []
               },
but it made no difference on 1.2.0-insiders3. I don't think I have understood what you've said about changing the settings. I guess I'll have to wait for #3123 to be fixed and see it is any better then.
I did an "ls -R" under Windows git bash in the VS Code terminal and it took about 55seconds. Obviously, that includes huge numbers of files that should be excluded (including many build artefacts).

**Comment by sean-mcmanus** [comment]
I created a test folder with symlinks and 1 source file and it ended up parsing the same file 20 times (i.e. it recursed into the symlink until the 20 depth was hit)...the Linux/Mac version correctly doesn't loop back to the same paths. I've filed an issue at https://github.com/microsoft/vscode-cpptools/issues/6864 .

**Comment by sean-mcmanus** [comment]
Hmm, I don't know why those settings aren't working unless you have a configurationProvider set. If the you're not hitting the symlink issue, then I'm not sure why we would be taking > 55 seconds to enumerate files.

I've switched to a breadth first search on Windows with a hardcoded 3 minute timeout per root path in browse.path, and I added logging messages when processing a depth takes > 2 seconds. The logging will show which paths are taking unusually long to process, so I'm hoping it would give more info on what is going wrong or if adding a setting would be good or not. This probably won't be released till 1.3.0-insiders though (not sure if files.exclude handling will be done by then).

The hardlinks would be the same as symlinks in regards to the looping behavior, but it would be hit if the hardlink pointed to a folder that was a parent.

**Comment by BillDenton** [comment]
As far as I can tell I'm not using a configurations provider:
![image](https://user-images.githubusercontent.com/17859340/106251950-90e5ac80-620d-11eb-9d45-a84421baadf2.png)

**Comment by sean-mcmanus** [comment]
Your screenshot just shows the user setting so it's possible configurationProvider could be set in the workspace settings or c_cpp_properties.json, but also running C/C++: Log Diagnostics should show...I'm assuming configurationProvider isn't set by those either, in which case I don't know yet why setting compilerPath and browse.path to empty would not work as expected.

Well, I started work on enabling the files.exclude checking to file enumeration and unifying the new Windows approach to Linux (it's a big change, so it might take a few weeks), so it looks like we could add some browse.enumerationDepthLimit and/or browse.enumerationTimeout settings; however, it seems like the logging around which folders are taking a lot of time to enumerate would enable users to exclude those folders, so the additional settings/timeouts might not be needed.

**Comment by sean-mcmanus** [comment]
With https://github.com/microsoft/vscode-cpptools/releases/tag/1.3.0-insiders5 we've fixed file enumeration so it will look at files.exclude and C_Cpp.files.exclude now and not generate a giant vector of all files at the start, which should reduce the max memory usage. That build should be close (or identical) to what we plan to ship next week as 1.3.0. Let us know if you're still hitting issues.
